Task: Deleting a statefulset

foxish opened the pull request:
This task explains deletion of a StatefulSet.

cc @erictune @kow3ns @enisoc @kubernetes/sig-apps

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/1737)
<!-- Reviewable:end -->

kargakis commented on docs/tasks/stateful-sets/deleting-a-statefulset.md:
```diff
@@ -0,0 +1,102 @@
+www-web-1   Bound     pvc-62d6750e-3822-11e6-b1b7-42010af00002   0                        21h
+
+$ kubectl delete pvc -l app=myapp
+$ kubectl get pv
```
Shouldn't this be `kubectl get pvc`?

foxish replied:
It is showing that the PVs are cleared due to the reclaim policy, when the PVCs are deleted.

janetkuo commented on _data/tasks.yml:
```diff
@@ -23,4 +23,7 @@ toc:
   section:
   - title: Assigning Pods to Nodes
     path: /docs/tasks/administer-cluster/assign-pods-nodes/
-
```
You need to update docs/tasks/index.md also

foxish replied:
@janetkuo Can we do it all at once, after we finish all the tasks? or do you think this PR should contain that change as well?

devin-donnelly replied:
Updating it as we go is going to cause merge conflicts, but updating it at the end is harder to coordinate. I think we should update it as we go and have the individual authors resolve the merge conflicts; it should just take a second and at least that way everything gets included.

janetkuo commented on docs/tasks/stateful-sets/deleting-a-statefulset.md:
```diff
@@ -0,0 +1,102 @@
+
+### Deleting a StatefulSet
+
+A Stateful Set can be deleted like other resources in kubernetes. 
```
Kubernetes

janetkuo commented on docs/tasks/stateful-sets/deleting-a-statefulset.md:
````diff
@@ -0,0 +1,102 @@
+
+```shell{% raw %}
+$ grace=$(kubectl get po web-0 --template '{{.spec.terminationGracePeriodSeconds}}')
+$ kubectl delete petset,po -l app=nginx
````
We don't need po here

enisoc commented on docs/tasks/stateful-sets/deleting-a-statefulset.md:
````diff
@@ -0,0 +1,102 @@
+
+```
+
+Deleting a Stateful Set through kubectl will scale it down to 0, thereby deleting all pods that are a part of it. If you wish to delete just the Stateful Set and not the pods, please use --cascade=false.
````
According to the [style guide](http://kubernetes.io/docs/contribute/style-guide/#use-simple-and-direct-language), we shouldn't say "please".

enisoc replied:
Also `--cascade=false` should be backtick quoted.

enisoc commented on docs/tasks/stateful-sets/deleting-a-statefulset.md:
````diff
@@ -0,0 +1,102 @@
+
+A Stateful Set can be deleted like other resources in kubernetes. 
+
+```shell
````
It seems a bit unnecessary for one-liners like this, but technically the [style guide](http://kubernetes.io/docs/contribute/style-guide/#separate-commands-from-output) wants us to keep shell commands separate from their output.

janetkuo replied:
Wow, thanks for pointing this out. I need to update my doc to match this guide also.

enisoc commented on docs/tasks/stateful-sets/deleting-a-statefulset.md:
````diff
@@ -0,0 +1,102 @@
+```shell{% raw %}
+$ grace=$(kubectl get po web-0 --template '{{.spec.terminationGracePeriodSeconds}}')
+$ kubectl delete petset,po -l app=nginx
+$ sleep $grace
````
Is this really safe? If so, maybe we should mention what this is for. Alternatively, just say in words to wait until a given `kubectl get` command returns no entries.

chrislovecnm replied:
This is safe :)

kargakis commented on docs/tasks/stateful-sets/deleting-a-statefulset.md:
```diff
@@ -0,0 +1,102 @@
+$ grace=$(kubectl get po web-0 --template '{{.spec.terminationGracePeriodSeconds}}')
+$ kubectl delete petset,po -l app=nginx
+$ sleep $grace
+$ kubectl delete pvc -l app=nginx
```
Is there a use-case for deleting the stateful set but leaving the pvc around? If not, shouldn't we make the statefulset reaper delete pvcs too?

chrislovecnm replied:
We are talking big data ;) There are plenty of use cases to keep the data around.

kargakis replied:
Right, we should call out explicitly in this example the implications of deleting the claims.

smarterclayton replied:
The default behavior is always preserve the PVC.  We *never* want to allow a user to accidentally delete a PVC (which is why it's not part of 'all').

Agree that in this case we should very clearly indicate what deleting the pvcs means.

janetkuo commented on docs/tasks/stateful-sets/deleting-a-statefulset.md:
```diff
@@ -0,0 +1,102 @@
+
+#### Persistent Volumes
+
+Deleting the pods will *not* delete the volumes. Until we finalize the recycle policy for these volumes they will have to get cleaned up by an admin. This is to ensure that you have the chance to copy data off the volume before deleting it. Simply deleting the PVC after the pods have left the [terminating state](/docs/user-guide/pods/index#termination-of-pods) should trigger deletion of the backi ...
```
Avoid using "we" according to style guide

enisoc replied:
Good catch. It also says not to use [passive voice](http://kubernetes.io/docs/contribute/style-guide/#use-active-voice), which makes it that much harder to avoid "we".

However, given that we also are not supposed to [make statements about the future](http://kubernetes.io/docs/contribute/style-guide/#avoid-statements-about-the-future), I would argue for just removing the entire sentence "Until we finalize ...".

chrislovecnm replied:
"The community" <- always works for me ;)

erictune replied:
+1 remove "until"

devin-donnelly commented on docs/tasks/stateful-sets/deleting-a-statefulset.md:
```diff
@@ -0,0 +1,82 @@
+
+{% capture prerequisites %}
+
+* If you are not using Stateful Sets, you can skip this task now. 
```
Not a good pre-requisite, though it is pretty Google-like.

How about: "This task assumes you have an application running on your cluster represented by a StatefulSet."

devin-donnelly commented on docs/tasks/stateful-sets/deleting-a-statefulset.md:
```diff
@@ -0,0 +1,82 @@
+
+### Deleting a StatefulSet
+
+A Stateful Set can be deleted like other resources in Kubernetes. 
```
Address the user directly; this is in passive voice.

Expound a bit: "You can delete a StatefulSet in the same way you delete other resources in kubernetes: use the `kubectl delete` command, and specify the StatefulSet either by file or by name."

devin-donnelly commented on docs/tasks/stateful-sets/deleting-a-statefulset.md:
````diff
@@ -0,0 +1,82 @@
+kubectl delete statefulsets <statefulset-name>
+```
+
+The associated headless service may need to be deleted separately after the Stateful Set itself is deleted.
````
Passive voice. "You may need to delete the associated headless service..."

devin-donnelly commented on docs/tasks/stateful-sets/deleting-a-statefulset.md:
````diff
@@ -0,0 +1,82 @@
+kubectl delete -f <file.yaml> --cascade=false
+```
+Setting `--cascade=false` causes the Pods managed by the Stateful Set to be left behind even after the Stateful Set object itself is deleted.
+If the pods have a label `app=myapp`, they can then be deleted as follows.
````
Passive voice. "They can then be deleted" -> "You can then delete them as follows:"

devin-donnelly commented on docs/tasks/stateful-sets/deleting-a-statefulset.md:
```diff
@@ -0,0 +1,82 @@
+
+#### Persistent Volumes
+
+Deleting the pods will *not* delete the volumes. This is to ensure that you have the chance to copy data off the volume before deleting it. Deleting the PVC after the pods have left the [terminating state](/docs/user-guide/pods/index#termination-of-pods) may trigger deletion of the backing Persistent Volumes depending on the storage class and reclaim policy. You should never assume ability to acc ...
```
"Deleting the Pods in a StatefulSet will *not* delete the associated volumes."

devin-donnelly replied:
"May" is for permission; use "might" instead.

devin-donnelly commented on docs/tasks/stateful-sets/deleting-a-statefulset.md:
```diff
@@ -0,0 +1,82 @@
+
+Deleting the pods will *not* delete the volumes. This is to ensure that you have the chance to copy data off the volume before deleting it. Deleting the PVC after the pods have left the [terminating state](/docs/user-guide/pods/index#termination-of-pods) may trigger deletion of the backing Persistent Volumes depending on the storage class and reclaim policy. You should never assume ability to acc ...
+
+**Note: you may lose all your data once the PVC is deleted, do this with caution.**
```
"Use caution when deleting a PVC, as it may lead to data loss."

devin-donnelly commented on docs/tasks/stateful-sets/deleting-a-statefulset.md:
```diff
@@ -0,0 +1,82 @@
+
+#### Complete deletion of a Stateful Set
+
+If you simply want to clean everything in a Stateful Set containing pods which have the label `app=myapp`:
```
By the way, I'm pretty sure the concept doc and other docs use "StatefulSet" with no space. Please conform to one style.

devin-donnelly commented on docs/tasks/stateful-sets/deleting-a-statefulset.md:
```diff
@@ -0,0 +1,82 @@
+**Note: you may lose all your data once the PVC is deleted, do this with caution.**
+
+#### Complete deletion of a Stateful Set
+
```
Sentence is too long and rolls in the example--separate it for clarity.

"To simply delete *everything* in a StatefulSet, including the associated pods, you can run a series of commands similar to the following:"

devin-donnelly commented on docs/tasks/stateful-sets/deleting-a-statefulset.md:
````diff
@@ -0,0 +1,82 @@
+sleep $grace
+kubectl delete pvc -l app=myapp
+```
+
````
"In the example above, the Pods have the label `app=myapp`; substitute your own label as appropriate."

devin-donnelly commented:
Not bad. Please address the docs comments and the merge conflict in the TOC.

foxish commented:
@devin-donnelly, Thanks! Addressed comments. PTAL.
Preview: https://deploy-preview-1737--kubernetes-io-vnext-staging.netlify.com/docs/tasks/manage-stateful-set/deleting-a-statefulset/

devin-donnelly commented:
Great for me. Does someone need to provide a tech LGTM?

enisoc commented:
LGTM if you're planning to replace the TODO with a link in a later PR.
